okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -721,8 +732,14 @@ void finishStream(
           frameWriter.rstStream(streamId, ErrorCode.CANCEL);
         }
         if (status != null) {
-          stream.transportState().transportReportStatus(status, stopDelivery,
-              trailers != null ? trailers : new Metadata());
+          boolean isCancelled =
+              (status.getCode() == Code.CANCELLED || status.getCode() == Code.DEADLINE_EXCEEDED);
+          stream
+              .transportState()
+              .transportReportStatus(
+                  status,
+                  isCancelled || stopDelivery,
```

Comment:
I'm not excited about the mixing of implicit (via Status) and explicit `stopDelivery`. It should be one way or the other, unless there was a _really good_ reason. In auditing the existing calls to this method, it seemed that only the rst stream one was depending on this behavior. If you'd like we can sit down and pair program/audit the callers.